common/src/main/python/dlpx/virtualization/common/util.py

```diff
@@ -0,0 +1,136 @@
+#
+# Copyright (c) 2021 by Delphix. All rights reserved.
+#
+
+
+"""
+Utility functions to convert between unicode and bytes.
+"""
+
+import six
+
+
+def to_bytes(string, encoding=None):
+    """
+    Converts the given object to binary object, bytes (Py3) or str (Py2).
+
+    :param string: The string like object to convert to bytes
+    :type string: ``object``
+    :param encoding: The encoding to encode the string with.
+    :type encoding: ``str``
+    :returns: The encoded string.
+    :rtype: ``bytes``
+    """
+    if string is None:
+        return
+
+    if not encoding:
+        encoding = "utf-8"
+
+    if isinstance(string, dict):
+        for k, v in string.items():
+            string[k] = to_bytes(v, encoding=encoding)
+        return string
+
+    if isinstance(string, list):
+        return [to_bytes(i, encoding=encoding) for i in string]
+
+    if isinstance(string, set):
+        return {to_bytes(i, encoding=encoding) for i in string}
+
+    if isinstance(string, str):
+        return _to_bytes(string, encoding)
+
+    return string
+
+
+def _to_bytes(string, encoding):
+    if six.PY3:
+        if isinstance(string, str):
+            return string.encode(encoding)
+        else:
+            return bytes(string)
+    else:
+        if isinstance(string, unicode):  # noqa
+            return string.encode(encoding)
+        else:
+            return str(string)
+
+
+def to_str(b, encodings=None):
```

Comment:
Is there a real-world use case in the SDK for saying "Here, try this list of encodings and then use whichever the first one is that doesn't cause a decode error"?

I could see this being useful for something like a text editor or a web browser, where neither the program nor the user has control/knowledge of what exactly the content is.  So, you'd want to guess at the encoding. You wouldn't always guess right, but you'd probably guess right "most of the time", and that's probably fine.

But, here, it seems like we'll probably have two cases?
  a) It's UTF-8 (e.g. any plugin-provided byte string)
  b) It's some other encoding, and the plugin has told us what encoding it is.

So, if I'm right about that, then I'd think we could avoid looping over lists of encodings here, and just always use one single encoding (which would default to UTF-8 but could optionally be specified by the caller)

Reply:
Yeah, I think you're right. I'll make this update as well